_posts/testing-wai.md

```diff
@@ -0,0 +1,222 @@
+
+```javascript
+const { bindings } = require('@dynamite-bud/rustfft');
+
+const {
+  Complex,
+  Algorithm,
+} = require('@dynamite-bud/rustfft/src/bindings/rustfft/rustfft.js');
+
+const { test, expect } = require('@jest/globals');
+
+test('dft test len 3', async () => {
+  const wasm = await bindings.rustfft();
+  let signal = [
+    { re: 1.0, im: 1.0 },
+    { re: 2.0, im: -3.0 },
+    { re: -1.0, im: 4.0 },
+  ];
+  let spectrum = [
+    { re: 2.0, im: 2.0 },
+    { re: -5.562177, im: -2.098076 },
+    { re: 6.562178, im: 3.09807 },
+  ];
+  let dft = Algorithm.newDft(wasm, signal.length, 'forward');
+  let output = dft.compute(signal);
+  output.forEach((element, index) => {
+    expect(element.re.toPrecision(5)).toBe(spectrum[index].re.toPrecision(5));
+  });
+});
+
+test('dft test len 4', async () => {
+  const wasm = await bindings.rustfft();
+  let signal = [
+    { re: 0.0, im: 1.0 },
+    { re: 2.5, im: -3.0 },
+    { re: -1.0, im: -1.0 },
+    { re: 4.0, im: 0.0 },
+  ];
+  let spectrum = [
+    { re: 5.5, im: -3.0 },
+    { re: -2.0, im: 3.5 },
+    { re: -7.5, im: 3.0 },
+    { re: 4.0, im: 0.5 },
+  ];
+  let dft = Algorithm.newDft(wasm, signal.length, 'forward');
+  let output = dft.compute(signal);
+  output.forEach((element, index) => {
+    expect(element.re.toPrecision(2)).toBe(spectrum[index].re.toPrecision(2));
+  });
+});
+```
+
+Running the `cargo test` now will automatically run the `jest` test for us.
+
+```shell-session
+Finished test [unoptimized + debuginfo] target(s) in 0.46s
+     Running unittests src/lib.rs (target/debug/deps/rustfft-e1e8db73703fde9b)
+
+running 0 tests
+
+test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
+
+     Running tests/rustfft-integration-tests.rs (target/debug/deps/rustfft_integration_tests-44dff04ad3dd20ad)
+yarn install v1.22.19
+warning package.json: No license field
+info No lockfile found.
+warning @dynamite-bud/rustfft@0.1.0: No license field
+[1/4] 🔍  Resolving packages...
+[2/4] 🚚  Fetching packages...
+[3/4] 🔗  Linking dependencies...
+[4/4] 🔨  Building fresh packages...
+
+success Saved lockfile.
+✨  Done in 0.07s.
+yarn install v1.22.19
+[1/4] 🔍  Resolving packages...
+success Already up-to-date.
+✨  Done in 0.08s.
+yarn run v1.22.19
+$ (...)/rustfft/node_modules/.bin/jest
+ PASS  ./rustfft.test.js
+  ✓ dft test len 3 (11 ms)
+  ✓ dft test len 4 (1 ms)
+
+Test Suites: 1 passed, 1 total
+Tests:       2 passed, 2 total
+Snapshots:   0 total
+Time:        0.169 s
+Ran all test suites.
+✨  Done in 0.76s.
+
+```
+
+We see here that our tests are ran using the `jest` testing suite and pass for both the test cases.
+
+## Conclusion
+
+In this tutorial, we got to know how to testing your published packages with WAI. The testing was done using JavaScript. This testing also concludes that our logic internally is working correctly and correct bindings are being generated for our package in JavaScript
```

Comment:
I asked ChatGPT to generate a conclusion for this article. Looks pretty neat.



In conclusion, testing your published packages with WAI using JavaScript is an effective way to ensure that your package's internal logic is working correctly and that the correct bindings are being generated in JavaScript. With the [wasmer-pack-testing](<https://crates.io/crates/wasmer-pack-testing>) automated testing framework, testing for WAI has become easier than ever. This framework allows for both logical checks using unit tests and code generation using snapshot testing. By following the steps outlined in this tutorial, you can harness the portability and logical computation power of WebAssembly and ensure the quality of your published packages.

Reply:
Sure thing!